modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -826,83 +826,130 @@ std::vector<InferenceEngine::InferRequest> cv::gimpl::ie::IECompiled::createInfe
-    size_t id = 0u;
-    m_idle_ids.pop(id);
+class AsyncInferExecutor : public IInferExecutor {
+public:
+    using IInferExecutor::IInferExecutor;
+    virtual void execute(const IInferExecutor::Task& task) override;
 
-    auto& request = m_requests[id];
+private:
+    void callback(Task task,
+                  IE::InferRequest request,
+                  IE::StatusCode code) noexcept;
+};
 
+void AsyncInferExecutor::execute(const IInferExecutor::Task& task) {
     using namespace std::placeholders;
     using callback_t = std::function<void(IE::InferRequest, IE::StatusCode)>;
-    request.SetCompletionCallback(
+    m_request.SetCompletionCallback(
             static_cast<callback_t>(
-                std::bind(&cv::gimpl::ie::RequestPool::callback, this,
-                          t, id, _1, _2)));
-    // NB: InferRequest is already marked as busy
-    // in case of exception need to return it back to the idle.
+                std::bind(&AsyncInferExecutor::callback, this, task, _1, _2)));
     try {
-        t.run(request);
+        task.set_input_data(m_request);
+        m_request.StartAsync();
     } catch (...) {
-        request.SetCompletionCallback([](){});
-        m_idle_ids.push(id);
+        m_request.SetCompletionCallback([](){});
+        m_notify();
         throw;
     }
 }
 
-void cv::gimpl::ie::RequestPool::callback(cv::gimpl::ie::RequestPool::Task task,
-                                          size_t id,
-                                          IE::InferRequest request,
-                                          IE::StatusCode code) noexcept {
-    // NB: Inference is over.
-    // 1. Run callback
-    // 2. Destroy callback to free resources.
-    // 3. Mark InferRequest as idle.
-    task.callback(request, code);
+void AsyncInferExecutor::callback(IInferExecutor::Task task,
+                                  IE::InferRequest     request,
+                                  IE::StatusCode       code) noexcept {
+    task.read_output_data(request, code);
     request.SetCompletionCallback([](){});
+    // NB: Notify pool that executor has finished.
+    m_notify();
+}
+
+class cv::gimpl::ie::RequestPool {
+public:
+
+    explicit RequestPool(std::vector<InferenceEngine::InferRequest>&& requests);
+
+    IInferExecutor::Ptr getIdleRequest();
+    void waitAll();
+
+private:
+    void setup();
+    void release(const size_t id);
+
+    QueueClass<size_t>               m_idle_ids;
+    std::vector<IInferExecutor::Ptr> m_requests;
+};
+
+void cv::gimpl::ie::RequestPool::release(const size_t id) {
     m_idle_ids.push(id);
 }
 
+// RequestPool implementation //////////////////////////////////////////////
+cv::gimpl::ie::RequestPool::RequestPool(std::vector<InferenceEngine::InferRequest>&& requests) {
+    for (size_t i = 0; i < requests.size(); ++i) {
+        m_requests.emplace_back(
+                std::make_shared<AsyncInferExecutor>(std::move(requests[0]),
```

Comment:
Here the question how to switch between `AsyncInferExecutor` and `SyncInferExecutor`. I see two options here:
1. Provide to user handle: `params.cfgInferenceAPI(ParamDesc::API api) // ParamDesc::API::ASYNC by default`
2. Calculate it based on number of infer requests. if nireq > 1 then it must be `AsyncInferRequest` since there is no sense to infer synchronously  with multiple infer requests.

I prefer the first option since it's more flexible and sometimes there is a difference between sync/async mode even with nireq == 1.

Reply:
What if we have different number of nireq each time - is it possible? Voting for the first option

Reply:
Updated, discussed locally